pkg/agent/agentclient/client.go

```diff
@@ -153,9 +155,13 @@ func (a *AgentClient) Serve(stopCh <-chan struct{}) {
 					}
 
 					close(dataCh)
+
+					a.connLock.Lock()
 					delete(a.connContext, connID)
+					a.connLock.Unlock()
 				},
 			}
+			a.connLock.Unlock()
```

Comment:
We need to get better about using defer.

Reply:
the only way is to have a method for every state change. I hear you. Let me address it later.